### ISSUE TITLE
Questions about names of weights in model definitions and in yaml files.

### ISSUE THREAD

**Issue by RizhaoCai** [NervanaSystems/distiller#341]
After checking the tutorial and FAQ, I have some questions about names of weights in both a model definition and in a yaml file.

Here are several questions about this.
------------------------------------------------------------------------------------

- Q1: Should all the weights to be pruned claimed in the yaml file?

    
     Let's take the yaml file "distiller/examples/pruning_filters_for_efficient_convnets/resnet56_cifar_channel_rank.yaml" for example. Between lines 120-130, it is stated as
![image](https://user-images.githubusercontent.com/37952703/62274023-970c9880-b471-11e9-81cb-2b1856061162.png)

There is no 'module.layer2.5.conv2.weight'. Does it mean this weight would not be pruned?
So I have to claim the weights that should be pruned in a yaml file? 

What is more, there are no weights of batch_norm layers stated. In the channel/filters pruning case, would the weights of batch_borm layers subsequent to the pruned layers be pruned automatically? 

---------------------------------------------------------------------------------------
- Q2: Should I correspond all weights names of a model definition and the yaml file?

Still the same example. There is "module.layer1.1.conv1.weight" at line 107 stated.
I instantiate a resnet56 model. 
Then I called `model.module.layer1.1` and I should fail because the SyntaxError
![image](https://user-images.githubusercontent.com/37952703/62274887-557ced00-b473-11e9-812f-de5f21de4cd0.png)

If I changed to `model.module.layer1[1].1`, I succeed.
![image](https://user-images.githubusercontent.com/37952703/62274985-82c99b00-b473-11e9-9aaa-96cd28f07488.png)

Therefore, when we are writing a yaml file and state the weights to be pruned, we should correspond to the names of the weights, right? And we should always use '.'  in the yaml file even though a module in the model definition counterpart should be retrieved by an index?

------------------------------------------------------

- Q3: If I don't use torch.nn.DataParallel to warp my model, I should remove 'module' in a yaml file?

When I create a resnet18,  I could not retrieve all the modules by calling `model.module`  if I don't use a torch.nn.DataParllel to warp it.
![image](https://user-images.githubusercontent.com/37952703/62275501-a93c0600-b474-11e9-84b2-6917ecfa3916.png)

In all the provided examples, the created models would be trained/pruned with multiple GPUs. In other words, they are warped by the torch.nn.DataParllel.

If I am going to prune my model with only one GPU, should I claim all the weights in a yaml file without 'module'?

Like this:
```
pruners:
  filter_pruner_70:
    class: 'L1RankedStructureParameterPruner'
    group_type: Channels
    desired_sparsity: 0.7
    group_dependency: Leader
    weights: [
      layer1.1.conv1.weight,
      layer1.0.conv1.weight,
      layer1.2.conv1.weight,
      layer1.3.conv1.weight,
      layer1.4.conv1.weight,
      layer1.5.conv1.weight,
      layer1.6.conv1.weight,
      layer1.7.conv1.weight,
      layer1.8.conv1.weight,
      layer2.0.conv1.weight,
      layer2.0.downsample.0.weight
      ]
 
```

I suppose I could also warp my model even if I only have one GPU so that I don't need to remove 'module'

I would like to discuss it because I want to make sure my guess in Q2

**Comment by levzlotnik** [comment]
Hi @RizhaoCai ,

Sorry for the delay, let's answer your questions one by one. TL;DR - you're correct on your guesses on all the questions 😄.

1.  The layers that aren't in the yaml file indeed won't be pruned. In the case you've showed - the yaml contains the `module.` prefix (which usually indicates that the model is `nn.DataParallel` or `nn.DistributedDataParallel`) and if your model doesn't contain these layers, they won't be pruned. That also goes for batchnorm layers, they aren't automatically pruned and hence they also need to be included in the file in the case you want to prune them.

2. In PyTorch, the submodules that are located in `nn.Sequential` or `nn.ModuleList` are accessed as in array/list (e.g. `model.sequential[0]` or `model.module_list[2]`) but their name in the graph is accessed through `.n` (e.g. `model.sequential.0` or `model.module_list.2`). This is why you get the syntax error, exactly as you described. The names of the weights are accessed exactly the same way, so you're correct, you should specify in the yaml file the *name* of parameter, and not the way you would access it in python.

3. You're correct - not using `nn.DataParallel` or `nn.DistributedDataParallel` won't wrap your model and hence there won't be a `module.` prefix. So you can choose one of 2 ways, either remove the prefix from the yaml or wrap the model in `nn.DataParallel`. I would remove the prefix so that pytorch wouldn't have to deal with `nn.DataParallel` shenanigans of splitting the data into one bunch before forward passing, and even though it takes a short amount of time it's still time you could be using to pass more tensors into the model :D. 

Cheers,
Lev

**Comment by RizhaoCai** [comment]
> Hi @RizhaoCai ,
> 
> Sorry for the delay, let's answer your questions one by one. TL;DR - you're correct on your guesses on all the questions 😄.
> 
> 1. The layers that aren't in the yaml file indeed won't be pruned. In the case you've showed - the yaml contains the `module.` prefix (which usually indicates that the model is `nn.DataParallel` or `nn.DistributedDataParallel`) and if your model doesn't contain these layers, they won't be pruned. BatchNormalization layers are _not_ pruned because all modern DNN compilers employ something called BN-folding in which the BN layers are removed (during graph inference execution) and the weights of a Convolutional layer preceding a BN layer are adjusted for the BN removal (BN layers scale and shift the feature-maps).
> 2. In PyTorch, the submodules that are located in `nn.Sequential` or `nn.ModuleList` are accessed as in array/list (e.g. `model.sequential[0]` or `model.module_list[2]`) but their name in the graph is accessed through `.n` (e.g. `model.sequential.0` or `model.module_list.2`). This is why you get the syntax error, exactly as you described. The names of the weights are accessed exactly the same way, so you're correct, you should specify in the yaml file the _name_ of parameter, and not the way you would access it in python.
> 3. You're correct - not using `nn.DataParallel` or `nn.DistributedDataParallel` won't wrap your model and hence there won't be a `module.` prefix. So you can choose one of 2 ways, either remove the prefix from the yaml or wrap the model in `nn.DataParallel`.
> 
> Cheers,
> Lev

I get clear now 😄
Thank you so much

**Comment by RizhaoCai** [comment]
> Hi @RizhaoCai ,
> 
> Sorry for the delay, let's answer your questions one by one. TL;DR - you're correct on your guesses on all the questions 😄.
> 
> 1. The layers that aren't in the yaml file indeed won't be pruned. In the case you've showed - the yaml contains the `module.` prefix (which usually indicates that the model is `nn.DataParallel` or `nn.DistributedDataParallel`) and if your model doesn't contain these layers, they won't be pruned. BatchNormalization layers are _not_ pruned because all modern DNN compilers employ something called BN-folding in which the BN layers are removed (during graph inference execution) and the weights of a Convolutional layer preceding a BN layer are adjusted for the BN removal (BN layers scale and shift the feature-maps).
> 2. In PyTorch, the submodules that are located in `nn.Sequential` or `nn.ModuleList` are accessed as in array/list (e.g. `model.sequential[0]` or `model.module_list[2]`) but their name in the graph is accessed through `.n` (e.g. `model.sequential.0` or `model.module_list.2`). This is why you get the syntax error, exactly as you described. The names of the weights are accessed exactly the same way, so you're correct, you should specify in the yaml file the _name_ of parameter, and not the way you would access it in python.
> 3. You're correct - not using `nn.DataParallel` or `nn.DistributedDataParallel` won't wrap your model and hence there won't be a `module.` prefix. So you can choose one of 2 ways, either remove the prefix from the yaml or wrap the model in `nn.DataParallel`.
> 
> Cheers,
> Lev

Hello Lev,

One more question:

Can I use wildcard character to specify the weights I would like to prune?

For example:

weights: [module.layer1.0.conv*.weight]

or 
weights: [module.layer1.*.conv1.weight]


Thanks
Riz
